docs/tutorial/t-spark-streaming.md

```diff
@@ -0,0 +1,335 @@
+zookeeper-k8s             active      1  zookeeper-k8s    3/edge    42  10.152.183.87   no       
+
+Unit                Workload  Agent  Address      Ports  Message
+data-integrator/0*  active    idle   10.1.29.186         
+kafka-k8s/0*        active    idle   10.1.29.184         
+kafka-test-app/0*   active    idle   10.1.29.185         Topic spark-streaming-store enabled with process producer
+zookeeper-k8s/0*    active    idle   10.1.29.182 
+```
+
+Now that `data-integrator` is deployed and integrated with `kafka-k8s`, we can get the credentials to connect to Kafka by running the `get-credentials` action exposed by the `data-integrator` charm.
+
+```bash
+juju run data-integrator/0 get-credentials
+```
+
+The output generated by this command is similar to the following:
+
+```
+Running operation 1 with 1 task
+  - task 2 on unit-data-integrator-0
+
+Waiting for task 2...
+kafka:
+  consumer-group-prefix: relation-9-
+  data: '{"extra-user-roles": "consumer,admin", "requested-secrets": "[\"username\",
+    \"password\", \"tls\", \"tls-ca\", \"uris\"]", "topic": "spark-streaming-store"}'
+  endpoints: kafka-k8s-0.kafka-k8s-endpoints:9092
+  password: g6c5gjg48IjTFld664ipkz8Khqb5FOG0
+  tls: disabled
+  topic: spark-streaming-store
+  username: relation-9
+  zookeeper-uris: zookeeper-k8s-0.zookeeper-k8s-endpoints:2181/kafka-k8s
+ok: "True"
+```
+
+As you can see, the endpoint, username and password to be used to authenticate with Kafka are displayed next to "endpoints", "username" and "password" respectively. It would be easier for us to store the username and password as variables so that they can be used later in the tutorial. For that, we can specify `--format=json` when running the `get-credentials` action, and then filter out username and password using `jq`.
+
+```bash
+KAFKA_USERNAME=$(juju run data-integrator/0 get-credentials --format=json | jq -r '.["data-integrator/0"].results.kafka.username')
+KAFKA_PASSWORD=$(juju run data-integrator/0 get-credentials --format=json | jq -r '.["data-integrator/0"].results.kafka.password')
+KAFKA_ENDPOINT=$(juju run data-integrator/0 get-credentials --format=json | jq -r '.["data-integrator/0"].results.kafka.endpoints')
+```
+
+For a moment, let's see the format of an event generated by `kafka-test-app`.
+
+```json
+{"timestamp": 1707128717.277745, "_id": "339e0f352b4843bfa0d512b796019a92", "origin": "kafka-test-app-0 (10.1.29.185)", "content": "Message #121"}
+```
+
+As we can see, the value of the "origin" key is the name and the IP address of the unit producing the events. We'll now write a Spark job in Python that counts the number of events grouped by the "origin" key in real time. 
+
+First of all, we need to authenticate with Kafka and load the events from the `spark-streaming-store` topic . This can be done using `spark.readStream` function as:
+
```

Comment:
When I got here I was a bit confused about how to follow this section - I started copying these python lines into a blank file but then realized that it was more of an explanation/overview of what the final python script would do. So I've reworded it in a way that I think makes it a bit more clear how to follow.
```suggestion

As we can see, the value of the "origin" key is the name and the IP address of the unit producing the events. 

Now we will write a Spark job in Python that counts the number of events grouped by the "origin" key in real time. We will first summarize the structure our script will take, and then provide the full python file for you to copy to your machine.

First, we will authenticate with Kafka and load the events from the `spark-streaming-store` topic . This can be done using the `spark.readStream` function as follows:
```